File: LeetCode/0005. Longest Palindromic Substring/Solution.cpp
/**
 * 94 / 94 test cases passed.
 * Status: Accepted
 * Runtime: 24 ms
 */
class Solution {
public:
    string longestPalindrome(string s) {
        int len = s.length();
        if (len < 2) return s;
        for (int i = 0; i < len; i++) {
            find(s, i, i);
            find(s, i, i + 1);
        }
        return s.substr(lo, max);
    }
private:
    void find(string s, int i, int j) {
        int len = s.length();
        while (i >= 0 && j < len && s[i] == s[j]) {
            i--;
            j++;
        }
        len = j - i - 1;
        if (len > max) {
            max = len;
            lo = i + 1;
        }
    }
    int max = 0;
    int lo = 0;
};

File: LeetCode/0005. Longest Palindromic Substring/solution.py
"""
180 / 180 test cases passed.
Runtime: 4420 ms
Memory Usage: 180 MB
"""
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        dp = [[False] * n for _ in range(n)]
        for i in range(n): dp[i][i] = True
        lo = hi = 0
        for i in range(n - 1, -1, -1):
            for j in range(i, n):
                if s[i] == s[j]:
                    if j - i <= 1:
                        dp[i][j] = True
                    else:
                        dp[i][j] = dp[i + 1][j - 1]
                if dp[i][j] and j - i > hi - lo:
                    lo, hi = i, j
        return s[lo:hi+1]
    